File: Problem31/main.java
import java.util.TreeSet;

public class main {
    public static  void main(String[] args) {
        TreeSet<Integer> denominations = new TreeSet<Integer>();
        int[] denominationInts = new int[] {1, 2, 5, 10, 20, 50, 100, 200};
        for (int denomination : denominationInts) {
            denominations.add(denomination);
        }
        System.out.println(ways(denominations, 200, denominations.last()));
    }

    private static int ways(TreeSet<Integer> denominations, int total, int max) {
        if (total == 0) {
            return 0;
        }
        
        int ways = 0;
        while (total > 0) {
            ways += (denominations, denominations.floor(total))
        }
        
        return ways;
    }
}

File: Problem31/main.hs
import Data.List

main :: IO ()
main = print $ ways [1, 2, 5, 10, 20, 50, 100, 200] 200

ways :: [Integer] -> Integer -> Integer
ways _ 0 = 1
ways denominations value = sum $
    [ways (filter (<= d) denominations) (value - d) |
        d <- filter (<= value) denominations]
